File: android/src/main/kotlin/com/tqc/ads/flutter_admob_native_ads/platform_view/NativeAdViewFactory.kt
package com.tqc.ads.flutter_admob_native_ads.platform_view

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/**
 * Factory for creating NativeAdPlatformView instances.
 */
class NativeAdViewFactory(
    private val messenger: BinaryMessenger,
    private val layoutType: String
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val creationParams = args as? Map<String, Any?> ?: emptyMap()

        return NativeAdPlatformView(
            context = context,
            viewId = viewId,
            creationParams = creationParams,
            messenger = messenger,
            layoutType = layoutType
        )
    }
}

File: ios/flutter_admob_native_ads/Sources/flutter_admob_native_ads/PlatformView/NativeAdViewFactory.swift
import Flutter
import UIKit

/// Factory for creating NativeAdPlatformView instances.
class NativeAdViewFactory: NSObject, FlutterPlatformViewFactory {

    private let messenger: FlutterBinaryMessenger
    private let layoutType: String

    init(messenger: FlutterBinaryMessenger, layoutType: String) {
        self.messenger = messenger
        self.layoutType = layoutType
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        let creationParams = args as? [String: Any] ?? [:]

        return NativeAdPlatformView(
            frame: frame,
            viewId: viewId,
            creationParams: creationParams,
            messenger: messenger,
            layoutType: layoutType
        )
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}
